### ISSUE TITLE
Download caches as non blocking service

### ISSUE THREAD

**Issue by ludoo0d0a** [cgeo/cgeo#330]
Could it be possible to move the download task as a non blocking service ?
Like some file manager, I suggest to add a "hide" button on the popup.

**Comment by okainov** [comment]
Was there any update on the implementation for this? Still would be really nice to have this functionality... @triakcz seemed to have proper implementation years ago, but seems like it was not merged for some reason...

**Comment by fm-sys** [comment]
If someone/I should ever work on this, a `IntentService` running is foreground mode should the way to go... (same as for download receiver)

**Comment by fm-sys** [comment]
Hi, coming back to this very old issue:

We meanwhile offer downloading the routing and map updates in the background, and this process could probably be reused for background cache updates without much additional effort.

However, one general question remains:
When should the updated data be presented to the user?
Currently, the user will see the results as soon as the download is complete and the progress bar gets dismissed. However the UI was blocked while the download was running anyway, so the question didn't arise. 

But how should cgeo behave when the refresh is happening in the background without blocking the UI? Refresh the UI for every single cache, refresh as soon as everything was downloaded or even don't refresh automatically at all to do not interrupt the user? 
The problem is, that a non-blocking download will allow users to navigate to totally different places in the meantime... What if a user views a cache (CacheDetailsActivity) which is refreshed at the same time? 

I would be interested to work on this issue as soon as these things are clarified... (hopefully time will permit)

**Comment by okainov** [comment]
>Currently, the user will see the results as soon as the download is complete and the progress bar gets dismissed. However the UI was blocked while the download was running anyway, so the question didn't arise.

Side note: if user just clicks outside of loading popup, it disappears and user has no idea about the status, if one starts new download, it doesn't work, only restarting c:geo works. So the problem was somehow still there.

As of "when to reload" - IMO at whole batch download finish, not after every cache (because downloads might be of hundreds/thousands of caches).

As a user I would expect there should be some activity/window/screen like "download status" which would show the status of downloads. Probably it might be one screen for all downloads in c:geo: maps, caches, whatever else. And when download finishes, there can be some kind of notification (starting with simple toast and up to actual notification and some "notification" overlay over the download manager icon)


>What if a user views a cache (CacheDetailsActivity) which is refreshed at the same time?

You mean, user downloads batch of caches, after starting the download, they open cache details? I suppose there should be some kind of lock/synchronization on the cache level behind the scenes, so when the process of refresh\download\send log\whatever starts for the given cache, no other activity/process can write to the very same cache. And as of download, if it sees that the cache was updated\downloaded already a few seconds ago, it might just skip the update (discussible?) or ignore it and still update.

**Comment by moving-bits** [comment]
Without diving into the other aspects mentioned in the two last post, there's one thing I like to add:

> We meanwhile offer downloading the routing and map updates in the background, and this process could probably be reused for background cache updates without much additional effort.

For downloading maps and routing data, we are using the system's built-in download service, whereas for caches we are using an http library, which we call with a couple of specific parameters. Not sure if this can be transferred to the download manager as well. Also, the resulting map/routing data files will be stored as physical files in some folder, whereas cache data needs to be analyzed internally. TMALSS: I'm unsure whether those two have enough in common (except that a background service may be involved for part of the processing).

**Comment by fm-sys** [comment]
> TMALSS: I'm unsure whether those two have enough in common (except that a background service may be involved for part of the processing).

For sure I'm aware of the technical background and we cannot simply use the system's built-in download service for caches. 

But we meanwhile have really promising experience with background services in general which has proven to work when it comes to importing the downloaded map files. 

And while implementing that background service back in the days, I already had created an abstraction layer having background cache downloads in mind ;-)

**Comment by eddiemuc** [comment]
The main new challenge which is hitting us when downloading caches in the background will be transaction handling, since then we have to build the whole system robust against parallel writing and reading of the database/cachecache. This might create a couple of funny side effects, we will see when implementation starts.

**Comment by fm-sys** [comment]
> we have to build the whole system robust against parallel writing and reading of the database/cachecache. This might create a couple of funny side effects

AFAIK the DB does only allow one read/write access at once. If multiple calls are made at the same time, they will be scheduled internally. But if `CacheCache` is thread-save as well, and whether this does really works as expected is another thing though...

**Comment by eddiemuc** [comment]
> AFAIK the DB does only allow one read/write access at once. If multiple calls are made at the same time, they will be scheduled internally. 

I fear the calls/threads trying to access the DB in parallel are then blocked until it is "free" again. If I remember correctly c:geo has a "one db connection/transaction only" usage pattern for the SQLite database - means: a single `SQLiteDatabase` instance is created, stored as quasi-singleton in `DataStore` and used for all accesses.

> But if `CacheCache` is thread-save as well, and whether this does really works as expected is another thing though...

Also it is not enough that `CacheCache` is thread-safe. All objects we read/write from/to it (mainly the Geocache object) needs to be thread-safe...

**Comment by fm-sys** [comment]
I created PR https://github.com/cgeo/cgeo/pull/12683 so we can test how it will behave in real life :)

> I fear the calls/threads trying to access the DB in parallel are then blocked until it is "free" again 

That's true. But as downloading caches is mostly time consuming cause of network traffic, the delay while we are actually writing to the DB should be rather short. 

However, the whole thing how `CacheCache`
does work internally is unfortunately a "Book with Seven Seals" for me... Maybe someone could step in if I encounter problems I'm unable to solve?

**Comment by fm-sys** [comment]
Just a short update, non-blocking cache downloads are now available for nightly users. The feature can be accessed via either via `Map > Background download` or `Lists > Manage Caches > Background refresh`.

Currently it's marked as experimental feature, so I would really like to get feedback about whether it's running smoothly... :)

**Comment by fm-sys** [comment]
The MVP is included in the nightlys since 3 months now. Is there anything what needs to be done before we can enable it for beta/production? 

A few ideas (not necessarily to be completed):
- Localization 
- Use it for storing single caches
- restrict download base on connection type (wifi/3g)
- notification download ETA
- icon in map to see status (instead top right gray disk, we could use a "double rounded arrow" sync symbol)




_Originally posted in https://github.com/cgeo/cgeo/issues/330#issuecomment-7421042 (modified & extended)_

**Comment by Lineflyer** [comment]
I have used it several times and least not noticed anything like data inconsistencies or crashed while in parallel browsing caches / using the app.

> icon in map to see status (instead top right gray disk, we could use a "double rounded arrow" sync symbol)

What do you mean by that? 
The download is visible in the system download manager, isn't it? Isn't this sufficient?

**Comment by okainov** [comment]
@fm-sys I think I missed this :)

Just tried today's nightly, first feedback:
- I miss indicator/button to see the status. Just saw the toast that download has completed, but I'd expect some icon on the map indeed (your last bullet, probably)
- Maps icons were not refreshed (no floppy disk got added) after download, neither refreshing after map moves or zooms helped, so I had to enter the cache and return to the map again to see the icons.
- Popup doesn't do anything on click (I'd expect to go to some status window where I can see which caches got downloaded/which are in progress)


But in general looks super nice, exactly what I was waiting for :) Thanks for efforts!

P.S. Seems like I haven't updated it for a while, now I see reworked Images tab with finally working description and possibility to zoom pictures directly in c:geo, updated map cursor, cool, good stuff :)

**Comment by fm-sys** [comment]
It's always difficult to develop good features if there is no user feedback. So thanks for testing and the constructive feedback, very appreciated! 

From what I read, it's currently not yet ready for a beta version. I will try to work on it a bit further to make the behavior more understandable for users.

Especially keeping the background service and the front-end in sync is quite a complex topic.

**Comment by fm-sys** [comment]
Yes, in the long-term we definitely don't want to maintain two ways for batch-downloading caches. The legacy code will be thrown out, as soon as the background download meets all wishes :) 

The recent past however has shown that it often makes sense to reimplement specific features from scratch, so they can run in parallel until the new one is considered to be good enough. At that point in time, we can simply make the new implementation visible for beta/release users and hide the old one in the same step.

We've done it this way with Filtering, Variables/Coordinate calculator and the new image tab. (...and a new unified map implementation is currently ongoing under the hood as well)

**Comment by fm-sys** [comment]
Coming back to the question about "store all / store unsaved" and/or a checkbox:

The old "store all" feature did **not** refresh all caches. Caches which where already downloaded were added to the selected list without additional refresh.
An additional refresh can now be triggered via "update already stored". 

So the checkbox and the two menu entries are two totally different features. However I would suggest to remove the "store unsaved" menu entry as it is IMHO not necessary anymore.
Meanwhile users can simply filter for `status > stored=no` and then use the "store all" (or "background download") feature.

**Comment by okainov** [comment]
As a user I have some lists (e.g. by countries) and I often have low zoom (high level)next to the borders. I have hundreds of caches visible.from each side of the borders but only one side was saved.before. I'd like to still have a way to "store unsaved" caches and don't update *lists* for the ones which were stored already.
Also I'm not really.interested in refresh tbh, If I need it, I do it separately.via "view as list - refresh all" or so.

**Comment by ztNFny** [comment]
How about a single menu item "Download caches" that then opens a dialog where you can select all those details instead of the current multiple menu items?

Store X caches to list XXX.
[ ] Add already downloaded caches to map
[ ] Refresh already downloaded caches
OK / Cancel

**Comment by ztNFny** [comment]
Another idea for this: Partial download of cache info. E.g. "listing only" to get status info or updated hint, "logs only" to load the recent logs.

**Comment by Lineflyer** [comment]
> Another idea for this: Partial download of cache info. E.g. "listing only" to get status info or updated hint, "logs only" to load the recent logs.

AFAIR such partial updates (also: pictures only for caches imported from GPX) are covered in some other issue around here.

**Comment by fm-sys** [comment]
PR #13084 improves the synchronization between background download and foreground activity...

**Comment by Lineflyer** [comment]
I think its the best time for it, as it forces all nightly users to use that new feature and possibly find hidden problems.
Up to now it seems to work good in my tests (although I have to admit that I only used it like 30% of times).

**Comment by ludoo0d0a** [comment]
11 years later, here we are  ! :) 

Well done !! Great job! 
I'm impatient to test it !

**Comment by fm-sys** [comment]
Glad to hear :)
